--- /home/ijntema/truenas/programming/securecopy/core-utils-based/V9.7.0/coreutils-9.7/src/cu-progs.mk.org	2025-08-14 14:54:55.265733000 +0200
+++ /home/ijntema/truenas/programming/securecopy/core-utils-based/V9.7.0/coreutils-9.7/src/cu-progs.mk	2025-08-14 14:54:55.287442300 +0200
@@ -105,10 +105,15 @@
 default__progs += src/tty
 default__progs += src/uname
 default__progs += src/unexpand
 default__progs += src/uniq
 default__progs += src/unlink
 default__progs += src/uptime
 default__progs += src/vdir
 default__progs += src/wc
 default__progs += src/whoami
 default__progs += src/yes
+# HIJN
+default__progs += src/crccp
+default__progs += src/crcsum
+default__progs += src/crcmv
+# END HIJN
